### PR TITLE
Add Model Explainer Server

### PR DESCRIPTION
**Pull request by cliveseldon**
**What this PR does / why we need it**:
Add a model explainer server needed as first step to allow users to get explanations for their deployed models.

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/kfserving/152)

### CODE REVIEW

File: python/alibiexplainer/alibiexplainer/explainer.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        self.protocol = protocol
+        self.method = method
+        self.training_data_uri = training_data_url
+        self.training_data = None
+        if self.method is ExplainerMethod.anchor_tabular:
+            self.explainer = AnchorTabular(self._predict_fn)
+        else:
+            raise NotImplementedError
+
+    def load(self):
```
Comment (ellis-bigelow) [inline python/alibiexplainer/alibiexplainer/explainer.py]: We need some sort of "loading" I think, to prevent repeat calls.

File: python/alibiexplainer/alibiexplainer/explainer.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+        logging.info("Loading explainer")
+        self.explainer.prepare()
+        self.ready = True
+
+    def _predict_fn(self, arr: np.ndarray) -> np.ndarray:
+        if self.protocol == "seldon.http":
```
Comment (ellis-bigelow) [inline python/alibiexplainer/alibiexplainer/explainer.py]: We should make these protocols an enum like you did above.

File: python/alibiexplainer/alibiexplainer/explainer.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+    def explain(self, inputs: List) -> Any:
+        if self.method is ExplainerMethod.anchor_tabular:
+            explaination = self.explainer.explain(inputs)
+            return json.loads(json.dumps(explaination, cls=NumpyEncoder))
+        else:
+            raise NotImplementedError
+
+
+class NumpyEncoder(json.JSONEncoder):
```
Comment (ellis-bigelow) [inline python/alibiexplainer/alibiexplainer/explainer.py]: Seems weird we need to encode this ourself. Can you elaborate? Is this something we should extract into a shared KFServing/py library?

File: python/alibiexplainer/alibiexplainer/explainer_method.py [hunk 4]
```diff
@@ -0,0 +1,9 @@
+from typing import List, Dict
+
+class ExplainerMethodImpl(object):
```
Comment (ellis-bigelow) [inline python/alibiexplainer/alibiexplainer/explainer_method.py]: What's the point of defining this interface?

File: python/alibiexplainer/setup.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+from setuptools import setup, find_packages
+
+tests_require = [
+    'pytest',
+    'pytest-tornasync',
+    'mypy'
+]
+
+setup(
```
Comment (ellis-bigelow) [inline python/alibiexplainer/setup.py]: Should we be exposing this as a pip package? My thought is that this implementation can be dockerized and made available on our GCR as a standard explainer implementation, but perhaps shouldn't be something people are installing themselves. What do you think?

File: python/alibiexplainer/Makefile [hunk 6]
```diff
@@ -0,0 +1,10 @@
+
```
Comment (ellis-bigelow) [inline python/alibiexplainer/Makefile]: Should this makefile live at the python level and have targets for all python packages we have?

File: python/kfserving/kfserving/protocols/seldon_http.py [hunk 7]
```diff
@@ -86,3 +89,5 @@ def wrap_response(self, response: List) -> Dict:
         ty = _get_request_ty(self.request)
         seldon_datadef = _create_seldon_data_def(arr, ty)
         return {"data": seldon_datadef}
+
+
```
Comment (ellis-bigelow) [inline python/kfserving/kfserving/protocols/seldon_http.py]: newlines?

File: python/kfserving/kfserving/explainer.py [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+class KFExplainer(object):
+
+    def __init__(self):
+        self.ready = False
+
+    def load(self):
+        raise NotImplementedError
+
+    def explain(self, inputs: List) -> Dict:
+        raise NotImplementedError
```
Comment (ellis-bigelow) [inline python/kfserving/kfserving/explainer.py]: nit: newline

File: python/kfserving/kfserving/model.py [hunk 9]
```diff
@@ -33,10 +33,6 @@ def predict(self, inputs: List) -> List:
     def postprocess(self, inputs: List) -> List:
         raise NotImplementedError
 
-    # TODO return type TBD
-    def explain(self, inputs: List) -> Any:
```
Comment (ellis-bigelow) [inline python/kfserving/kfserving/model.py]: I have some thoughts on this :)

File: python/kfserving/kfserving/server.py [hunk 10]
```diff
@@ -53,6 +54,7 @@ def __init__(self, protocol: str = args.protocol, http_port: int = args.http_por
         self.grpc_port = grpc_port
         self.protocol = protocol
         self._http_server: Optional[tornado.httpserver.HTTPServer] = None
+        self.explainer = None
```
Comment (ellis-bigelow) [inline python/kfserving/kfserving/server.py]: This feels a little awkward. Can you share your thoughts on encoding these pieces into the KFServer vs wiring it all through the model object?

File: python/kfserving/kfserving/server.py [hunk 11]
```diff
@@ -91,17 +96,43 @@ def register_model(self, model: KFModel):
         self.registered_models[model.name] = model
 
 
+def _getRequestHandler(protocol, request: Dict) -> RequestHandler:
```
Comment (ellis-bigelow) [inline python/kfserving/kfserving/server.py]: camel -> snake?

File: python/.gitignore [hunk 12]
```diff
@@ -3,3 +3,4 @@ build
 .coverage
 *.egg-info
 __pycache__
+*.joblib
```
[the remark this answers is not in the document]
Comment (cliveseldon) [reply python/.gitignore]: I assumed we wouldn't be storing binary joblibs in the repo but happy to remove.

File: python/alibiexplainer/Makefile [hunk 13]
```diff
@@ -0,0 +1,10 @@
...
+dev_install:
+	pip install -e .
+	pip install -e .[test]
```
[the remark this answers is not in the document]
Comment (cliveseldon) [reply python/alibiexplainer/Makefile]: Looks like just `pip install -e .[test]` is enough to do both

File: python/alibiexplainer/README.md [hunk 14]
```diff
@@ -0,0 +1,206 @@
...
+python3 -m alibiexplainer
+usage: alibiexplainer [-h] [--http_port HTTP_PORT] [--grpc_port GRPC_PORT]
+                      [--protocol {tensorflow.http,seldon.http}] --model_url
+		      MODEL_URL [--method {ExplainerMethod.anchor_tabular}]
+alibiexplainer: error: the following arguments are required: --model_url
+```
+
+## Samples
+
+To run a local example follow the [income classifier explanation sample](./samples/income/README.md).
```
[the remark this answers is not in the document]
Comment (cliveseldon) [reply python/alibiexplainer/README.md]: Will move to top level samples.

File: python/alibiexplainer/alibiexplainer/anchor_tabular.py [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+
+class AnchorTabular(ExplainerMethodImpl):
+
+    def __init__(self, predict_fn: Callable):
+        self.predict_fn = predict_fn
+        self.cmap = None
+        self.anchors_tabular: Optional[alibi.explainers.AnchorTabular] = None
+
+    def prepare(self):
+        training_data_url = os.environ.get("TRAINING_DATA_URL")
```
[the remark this answers is not in the document]
Comment (cliveseldon) [reply python/alibiexplainer/alibiexplainer/anchor_tabular.py]: This is a big question. Some explanation algorithms need (or work better) if you have the training data. So I suppose my rationale is if it can be specified in the resource spec then probably should be an argument. Its probably important enough to be there so will add as optional argument with validation based on method.

File: python/alibiexplainer/samples/income/README.md [hunk 16]
```diff
@@ -0,0 +1,172 @@
...
+          [19, "Private", "High School grad", "Never-Married", "Service", "Own-child", "White", "Female", 0, 0, 15, "United-States"],
+          [38, "Private", "High School grad", "Never-Married", "Blue-Collar", "Husband", "White", "Male", 0, 0, 40, "United-States"],
+          [58, "Private", "Dropout", "Never-Married", "Service", "Husband", "White", "Male", 0, 0, 20, "United-States"],
+          [47, "Private", "High School grad", "Never-Married", "Professional", "Husband", "White", "Male", 0, 0, 40, "British-Commonwealth"],
+          [26, "Private", "Bachelors", "Never-Married", "Professional", "Not-in-family", "White", "Male", 0, 0, 38, "United-States"],
+          [21, "Private", "High School grad", "Never-Married", "Blue-Collar", "Own-child", "White", "Male", 0, 0, 12, "United-States"],
+          [43, "Private", "Dropout", "Never-Married", "Blue-Collar", "Wife", "Asian-Pac-Islander", "Female", 0, 0, 40, "SE-Asia"],
+          [39, "Private", "Bachelors", "Never-Married", "Admin", "Husband", "White", "Male", 0, 0, 40, "United-States"]
+        ],
+        covered_false: [
```
[the remark this answers is not in the document]
Comment (cliveseldon) [reply python/alibiexplainer/samples/income/README.md]: Fixed

File: python/alibiexplainer/samples/income/README.md [hunk 17]
```diff
@@ -0,0 +1,172 @@
...
+Now, run a KFServing sklearn server with this model:
+
+```
+python -m sklearnserver --model_dir ./  --model_name income --protocol seldon.http
+```
+
+In a different terminal start the Alibi Explainer:
+
+```
+TRAINING_DATA_URL=./train.joblib FEATURE_NAMES_URL=./features.joblib CATEGORICAL_MAP_URL=./category_map.joblib python -m alibiexplainer --model_url http://localhost:8080/models/income:predict --protocol seldon.http --http_port 8081
```
[the remark this answers is not in the document]
Comment (cliveseldon) [reply python/alibiexplainer/samples/income/README.md]: Yes, until we fix our path structure so model name is not passed in we will require the full URL. See issue #144

File: python/kfserving/kfserving/server.py [hunk 18]
```diff
@@ -91,17 +96,43 @@ def register_model(self, model: KFModel):
...
+
+class ModelExplainHandler(tornado.web.RequestHandler):
+
+    def initialize(self, protocol: str, explainer: KFExplainer):
+        self.protocol = protocol
+        self.explainer = explainer
+
+    def post(self):
+
+        if not self.explainer.ready:
```
[the remark this answers is not in the document]
Comment (cliveseldon) [reply python/kfserving/kfserving/server.py]: I think each thread will call `load` separately which might causes issues if we are downloading files. Maybe we should open a separate issue for this as it applies to predict as well?

File: python/alibiexplainer/alibiexplainer/explainer.py (hunk 1 above)
Comment (cliveseldon) [reply python/alibiexplainer/alibiexplainer/explainer.py]: I think we can just use `self.ready` as tornado would be single threaded

File: python/alibiexplainer/alibiexplainer/explainer.py (hunk 3 above)
Comment (cliveseldon) [reply python/alibiexplainer/alibiexplainer/explainer.py]: Numpy types can't be serialized by default by the json library so you need to convert to python types.
Will move to kfserving util.

File: python/alibiexplainer/alibiexplainer/explainer_method.py (hunk 4 above)
Comment (cliveseldon) [reply python/alibiexplainer/alibiexplainer/explainer_method.py]: So future different explainer implementations can be specified. At present there is just AnchorsTabular taken from Alibi.

File: python/alibiexplainer/setup.py (hunk 5 above)
Comment (cliveseldon) [reply python/alibiexplainer/setup.py]: I agree it will be mainly used as a provided image. I suppose that is true for the xgbserver and skserver as well which also have `setup.py`. For ease of development its use but I agree they probably won't be published to PyPi.

File: python/kfserving/kfserving/server.py (hunk 10 above)
Comment (cliveseldon) [reply python/kfserving/kfserving/server.py]: I'll change it so its back in `KFModel`. I think further refactoring could be done in a further PR.

**Comment by ellis-bigelow** [comment]
Awesome work Clive! Really excited to see customer feedback on this.
